README.md
# vscode-translate

[![Current Version](https://vsmarketplacebadge.apphb.com/version/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)
[![Install Count](https://vsmarketplacebadge.apphb.com/installs/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)
[![Open Issues](https://vsmarketplacebadge.apphb.com/rating/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)

## 语言翻译

按 CTRL + t to 打开翻译 then u can select the text and see the change of statusBar.
press CTRL + t to turn off the function.

### 似乎

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/translate_show.gif)

* it will read http.proxy at user-settings to get data by proxy

* and you can set the user-settings to change from language and target language

* as for translation.api only google and baidu provided (default is baidu)

* translation.languageDetection = true means do detection before language so that we could translate documents between from language and target language. (default is true)

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/config.png)

### 按住 CTRL+SHIFT+R 替换选择文本为翻译结果

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/replace_show.gif)

## 语言列表 (来源 google)

你可以看 [https://cloud.google.com/translate/docs/languages](https://cloud.google.com/translate/docs/languages)

| Language        | ISO-639-1 Code
| ------------- |:-------------: |
| Afrikaans | af |
| Albanian | sq |
| Amharic | am |
| Arabic | ar |
| Armenian | hy |
| Azeerbaijani | az |
| Basque | eu |
| Belarusian | be |
| Bengali | bn |
| Bosnian | bs |
| Bulgarian | bg |
| Catalan | ca |
| Cebuano | ceb (ISO-639-2) |
| Chichewa | ny |
| Chinese (Simplified) | zh-CN (BCP-47) |
| Chinese (Traditional) | zh-TW (BCP-47) |
| Corsican | co |
| Croatian | hr |
| Czech | cs |
| Danish | da |
| Dutch | nl |
| English | en |
| Esperanto | eo |
| Estonian | et |
| Filipino | tl |
| Finnish | fi |
| French | fr |
| Frisian | fy |
| Galician | gl |
| Georgian | ka |
| German | de |
| Greek | el |
| Gujarati | gu |
| Haitian Creole | ht |
| Hausa | ha |
| Hawaiian | haw (ISO-639-2) |
| Hebrew | iw |
| Hindi | hi |
| Hmong | hmn (ISO-639-2) |
| Hungarian | hu |
| Icelandic | is |
| Igbo | ig |
| Indonesian | id |
| Irish | ga |
| Italian | it |
| Japanese | ja |
| Javanese | jw |
| Kannada | kn |
| Kazakh | kk |
| Khmer | km |
| Korean | ko |
| Kurdish | ku |
| Kyrgyz | ky |
| Lao | lo |
| Latin | la |
| Latvian | lv |
| Lithuanian | lt |
| Luxembourgish | lb |
| Macedonian | mk |
| Malagasy | mg |
| Malay | ms |
| Malayalam | ml |
| Maltese | mt |
| Maori | mi |
| Marathi | mr |
| Mongolian | mn |
| Burmese | my |
| Nepali | ne |
| Norwegian | no |
| Pashto | ps |
| Persian | fa |
| Polish | pl |
| Portuguese | pt |
| Punjabi | ma |
| Romanian | ro |
| Russian | ru |
| Samoan | sm |
| Scots Gaelic | gd |
| Serbian | sr |
| Sesotho | st |
| Shona | sn |
| Sindhi | sd |
| Sinhala | si |
| Slovak | sk |
| Slovenian | sl |
| Somali | so |
| Spanish | es |
| Sundanese | su |
| Swahili | sw |
| Swedish | sv |
| Tajik | tg |
| Tamil | ta |
| Telugu | te |
| Thai | th |
| Turkish | tr |
| Ukrainian | uk |
| Urdu | ur |
| Uzbek | uz |
| Vietnamese | vi |
| Welsh | cy |
| Xhosa | xh |
| Yiddish | yi |
| Yoruba | yo |
| Zulu | zu |

CHANGELOG.md
# 更新日志

## 支线版本

作者好久没更新了，准备改成默认谷歌


out/src/extension.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode_1 = require("vscode");
const WebRequest = require("web-request");
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
let isActive = false, translateResult = '', google2baidu = {
    en: 'en',
    th: 'th',
    ru: 'ru',
    pt: 'pt',
    el: 'el',
    nl: 'nl',
    pl: 'pl',
    bg: 'bul',
    et: 'est',
    da: 'dan',
    fi: 'fin',
    cs: 'cs',
    ro: 'rom',
    sl: 'slo',
    sv: 'swe',
    hu: 'hu',
    de: 'de',
    it: 'it',
    zh: 'zh',
    'zh-CN': 'zh',
    'zh-TW': 'cht',
    'zh-HK': 'yue',
    ja: 'jp',
    ko: 'kor',
    es: 'spa',
    fr: 'fra',
    ar: 'ara'
};
function activate(context) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    context.subscriptions.push(new Translate());
    context.subscriptions.push(vscode_1.commands.registerCommand('extension.translate', () => {
        if (!isActive) {
            vscode_1.window.showInformationMessage('打开翻译');
            isActive = true;
        }
        else {
            vscode_1.window.showInformationMessage('关闭翻译');
            isActive = false;
        }
    }));
    context.subscriptions.push(vscode_1.commands.registerCommand('extension.translateReplace', () => {
        if (!isActive) {
            return;
        }
        var editor = vscode_1.window.activeTextEditor, selection = editor.selection;
        editor.edit(edit => edit.replace(selection, translateResult));
    }));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;
class Translate {
    constructor() {
        if (!this.statusBarItem) {
            this.statusBarItem = vscode_1.window.createStatusBarItem(vscode_1.StatusBarAlignment.Left);
        }
        let subscriptions = [];
        vscode_1.window.onDidChangeTextEditorSelection(this.updateTranslate, this, subscriptions);
        this.disposable = vscode_1.Disposable.from(...subscriptions);
        this.updateTranslate();
        this.statusBarItem.show();
    }
    updateTranslate() {
        var cfg = vscode_1.workspace.getConfiguration(), proxy = String(cfg.get("http.proxy")), api = String(cfg.get("translation.api")), targetLanguage = String(cfg.get("translation.targetLanguage")), fromLanguage = String(cfg.get("translation.fromLanguage")), languageDetection = Boolean(cfg.get("translation.languageDetection"));
        let editor = vscode_1.window.activeTextEditor;
        if (!isActive || !editor) {
            this.statusBarItem.hide();
            return;
        }
        let doc = editor.document, str = doc.getText(editor.selection).trim();
        if (str.trim() == '')
            return;
        setTimeout(() => {
            languageDetection
                ? this.languageDetection(str, fromLanguage).then((isReverse) => {
                    if (isReverse) {
                        [fromLanguage, targetLanguage] = [targetLanguage, fromLanguage];
                    }
                    this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
                })
                : this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
        }, 1000);
    }
    dotranslate(str, proxy, api, targetLanguage, fromLanguage) {
        var statusBarItem = this.statusBarItem;
        var translateStr = api == 'baidu' ? this.baiduTranslate(str, google2baidu[targetLanguage], google2baidu[fromLanguage]) : this.googleTranslate(str, targetLanguage, fromLanguage);
        WebRequest.get(translateStr, { "proxy": proxy }).then(function (TResult) {
            var rs = '', res = JSON.parse(TResult.content.toString());
            if (api == 'baidu') {
                if (res.error)
                    return;
                rs = res.trans_result.data[0].dst;
            }
            else {
                var result = [];
                res.sentences.forEach(function (v) {
                    result.push(v.trans);
                });
                rs = result.join(',');
            }
            statusBarItem.text = translateResult = rs;
            statusBarItem.show();
        });
    }
    languageDetection(str, fromLanguage) {
        return WebRequest.post('http://fanyi.baidu.com/langdetect', { "formData": { "query": str }, "timeout": 500 }).then(function (result) {
            var res = JSON.parse(result.content);
            if (res.error || res.lan == fromLanguage || fromLanguage == '') {
                return false;
            }
            else {
                return true;
            }
        }, function () {
            return false;
        });
    }
    baiduTranslate(str, targetLanguage, fromLanguage) {
        return 'http://fanyi.baidu.com/v2transapi?query=' + str + (fromLanguage ? '&from=' + fromLanguage : '') + (targetLanguage ? '&to=' + targetLanguage : '');
    }
    googleTranslate(str, targetLanguage, fromLanguage) {
        return 'https://translate.google.cn/translate_a/single?client=gtx&sl=' + (fromLanguage || 'auto') + '&tl=' + (targetLanguage || 'auto') + '&hl=zh-CN&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon&q=' + str;
    }
    dispose() {
        this.statusBarItem.dispose();
        this.disposable.dispose();
    }
}
//# sourceMappingURL=extension.js.map

src/extension.ts
'use strict';
// 模块“vscode”包含VS代码扩展性API
// 导入模块并在下面的代码中使用别名vscode引用它
import {
    window,
    workspace,
    commands,
    Disposable,
    ExtensionContext,
    StatusBarAlignment,
    StatusBarItem,
    TextDocument
} from 'vscode';
import * as WebRequest from 'web-request';
// 当你的扩展被激活时，这个方法被调用
// 你的扩展在命令执行的第一次被激活
let isActive = false;
// 翻译结果
let translateResult = '';
// 将谷歌语言码转换为百度语言码
const google2baidu = {
        en: 'en',
        th: 'th',
        ru: 'ru',
        pt: 'pt',
        el: 'el',
        nl: 'nl',
        pl: 'pl',
        bg: 'bul',
        et: 'est',
        da: 'dan',
        fi: 'fin',
        cs: 'cs',
        ro: 'rom',
        sl: 'slo',
        sv: 'swe',
        hu: 'hu',
        de: 'de',
        it: 'it',
        zh: 'zh',
        'zh-CN': 'zh',
        'zh-TW': 'cht',
        'zh-HK': 'yue',
        ja: 'jp',
        ko: 'kor',
        es: 'spa',
        fr: 'fra',
        ar: 'ara'
    };
/**
 * 激活函数
 * @param context
 */
export function activate(context: ExtensionContext) {
    // 使用控制台输出诊断信息（console.log）和错误（console.error）
    // 当你的插件被激活时，这行代码将只执行一次
    context.subscriptions.push(new Translate());
    // 插入翻译指令
    context.subscriptions.push(
        commands.registerCommand('extension.translate', () => {
            if (!isActive) {
                window.showInformationMessage('打开翻译');
                isActive = true;
            } else {
                window.showInformationMessage('关闭翻译');
                isActive = false;
            }
        })
    );
    // 插入替换指令
    context.subscriptions.push(
        commands.registerCommand('extension.translateReplace', () => {
            if (!isActive) { return; }
            let editor = window.activeTextEditor;
            let selection = editor.selection;
            editor.edit(edit => edit.replace(selection, translateResult));
        })
    );
}

// 当您的插件被停用时，此方法被调用
export function deactivate() { }

/**
 * 翻译类
 */
class Translate {

    private statusBarItem: StatusBarItem;
    private disposable: Disposable;
    /**
     * 构建函数
     */
    constructor() {
        if (!this.statusBarItem) {
            this.statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
        }
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(this.updateTranslate, this, subscriptions);
        this.disposable = Disposable.from(...subscriptions);
        this.updateTranslate();
        this.statusBarItem.show();
    }
    /**
     * 更新翻译
     */
    public updateTranslate() {
        const cfg = workspace.getConfiguration();
        let proxy = String(cfg.get("http.proxy"));
        let api = String(cfg.get("translation.api"));
        let targetLanguage = String(cfg.get("translation.targetLanguage"));
        let fromLanguage = String(cfg.get("translation.fromLanguage"));
        let languageDetection = Boolean(cfg.get("translation.languageDetection"));
        let editor = window.activeTextEditor;
        if (!isActive || !editor) {
            return this.statusBarItem.hide();
        }
        let doc = editor.document;
        let str = doc.getText(editor.selection).trim();
        if (str.trim() == '') return;
        setTimeout(() => {
            if (languageDetection) {
                this.languageDetection(str, fromLanguage).then((isReverse) => {
                    if (isReverse) {
                        [fromLanguage, targetLanguage] = [targetLanguage, fromLanguage];
                    }
                    this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
                })
            } else {
                this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
            }
        }, 1000);
    }
    /**
     * 翻译函数
     * * 调用接口，返回翻译结果，写入状态栏
     * @param str 内容
     * @param proxy 代理
     * @param api 接口 默认百度 也可以是谷歌
     * @param targetLanguage 目标语言
     * @param fromLanguage 源语言
     */
    private dotranslate(str, proxy, api, targetLanguage, fromLanguage) {
        let statusBarItem = this.statusBarItem;
        let translateStr = this.baiduTranslate(str, google2baidu[targetLanguage], google2baidu[fromLanguage])
        if (api == 'google'){
            translateStr = this.googleTranslate(str, targetLanguage, fromLanguage)
        }
        WebRequest.get(translateStr, { "proxy": proxy }).then(function (TResult) {
            let rs = '';
            let res = JSON.parse(TResult.content.toString());
            if (api == 'baidu') {
                if (res.error) return;
                rs = res.trans_result.data[0].dst;
            } else {
                var result = []
                res.sentences.forEach(v => {
                    result.push(v.trans)
                })
                rs = result.join(',');
            }
            statusBarItem.text = translateResult = rs;
            statusBarItem.show();
        });
    }
    /**
     * 语言检测
     * @param str 文字
     * @param fromLanguage 源语言
     */
    private languageDetection(str, fromLanguage) {
        return WebRequest.post('http://fanyi.baidu.com/langdetect',
            {
                "formData": { "query": str }, "timeout": 500
            }
        ).then(function (result) {
                var res = JSON.parse(result.content);
                if (res.error || res.lan == fromLanguage || fromLanguage == '') {
                    return false;
                } else {
                    return true;
                }
            },
            function () {
                return false;
            }
        );
    }
    /**
     * 百度翻译接口
     */
    private baiduTranslate(str, targetLanguage, fromLanguage) {
        return `http://fanyi.baidu.com/v2transapi?query=${str}&from=${fromLanguage}&to=${targetLanguage}`;
    }
    /**
     * 谷歌翻译接口
     */
    private googleTranslate(str, targetLanguage, fromLanguage) {
        fromLanguage = fromLanguage || 'auto';
        targetLanguage = targetLanguage || 'auto';
        return `https://translate.google.cn/translate_a/single?client=gtx&sl=${fromLanguage}&tl=${targetLanguage}&hl=zh-CN&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon&q=${str}`;
    }
    /**
     * 处置
     */
    dispose() {
        this.statusBarItem.dispose();
        this.disposable.dispose();
    }
}
